src/decoder/flac.rs

```diff
@@ -145,27 +611,130 @@ where
+    /// - `upper_bound`: Maximum number of samples that might be available (None if unknown)
+    ///
+    /// # Accuracy
+    ///
+    /// - **With metadata**: Exact remaining sample count (lower == upper)
+    /// - **Without metadata**: Conservative estimate based on current block
+    /// - **Stream exhausted**: (0, Some(0))
+    ///
+    /// # Implementation
+    ///
+    /// The lower bound counts buffered samples that are immediately available.
+    /// The upper bound uses total stream metadata when available for precise counting.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // Samples already decoded and buffered (guaranteed available)
+        let buffered_samples = self
+            .current_span_len()
+            .unwrap_or(0)
+            .saturating_sub(self.current_block_off);
+
+        if let Some(total_samples) = self.total_samples {
+            let total_remaining = total_samples.saturating_sub(self.samples_read) as usize;
+            (buffered_samples, Some(total_remaining))
+        } else if self.current_block.is_empty() {
+            // Stream exhausted (no more blocks available)
+            (0, Some(0))
+        } else {
+            (buffered_samples, None)
+        }
+    }
 }
 
-/// Returns true if the stream contains FLAC data, then tries to rewind it to where it was.
-fn is_flac<R>(mut data: R) -> bool
+/// Probes input data to detect FLAC format.
+///
+/// This function attempts to parse the FLAC magic bytes and stream info header to determine if the
+/// data contains a valid FLAC stream. The stream position is restored regardless of the result.
+///
+/// # Arguments
+///
+/// * `data` - Mutable reference to the input stream to probe
+///
+/// # Returns
+///
+/// - `true` if the data appears to contain a valid FLAC stream
+/// - `false` if the data is not FLAC or is corrupted
+///
+/// # Implementation
+///
+/// Uses the common `utils::probe_format` helper which:
+/// 1. Saves the current stream position
+/// 2. Attempts FLAC detection using `claxon::FlacReader`
+/// 3. Restores the original stream position
+/// 4. Returns the detection result
+///
+/// # Performance
+///
+/// This function only reads the minimum amount of data needed to identify
+/// the FLAC format (magic bytes and basic header), making it efficient for
+/// format detection in multi-format scenarios.
+fn is_flac<R>(data: &mut R) -> bool
 where
     R: Read + Seek,
 {
-    let stream_pos = data.stream_position().unwrap_or_default();
-    let result = FlacReader::new(data.by_ref()).is_ok();
-    let _ = data.seek(SeekFrom::Start(stream_pos));
-    result
+    utils::probe_format(data, |reader| FlacReader::new(reader).is_ok())
+}
+
+/// Converts claxon decoder errors to rodio seek errors.
+///
+/// This implementation provides error context preservation when FLAC decoding operations fail
+/// during seeking. The original `claxon` error is wrapped in an `Arc` for thread safety and
```

Comment:
Implementation detail not relevant to API user. Viewable in standard way in the generated docs.

<img width="2084" height="1910" alt="image" src="https://github.com/user-attachments/assets/b1c0ef09-5488-44bd-9418-4a57ce89b3e6" />